File: on_audio_query/android/src/main/kotlin/com/lucasjosino/on_audio_query/methods/observers/SongsObserver.kt
package com.lucasjosino.on_audio_query.query.observers

import android.content.Context
import android.database.ContentObserver
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.provider.MediaStore
import com.lucasjosino.on_audio_query.query.SongsQuery
import io.flutter.plugin.common.EventChannel

class SongsObserver(
    private val context: Context,
) : ContentObserver(Handler(Looper.getMainLooper())), EventChannel.StreamHandler {

    companion object {
        //
        private val URI: Uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
    }

    val isRunning: Boolean get() = pIsRunning

    //
    private val query: SongsQuery = SongsQuery()

    //
    private var sink: EventChannel.EventSink? = null
    private var args: Map<*, *>? = null

    //
    private var pIsRunning: Boolean = false

    override fun onChange(selfChange: Boolean) {
        //
        query.init(context, sink = sink, args = args)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        //
        sink = events
        args = arguments as Map<*, *>

        //
        context.contentResolver.registerContentObserver(URI, true, this)

        //
        pIsRunning = true

        // Send the initial result.
        query.init(context, sink = sink, args = args)
    }

    override fun onCancel(arguments: Any?) {
        //
        context.contentResolver.unregisterContentObserver(this)

        //
        pIsRunning = false

        //
        sink = null
    }
}

File: on_audio_query/ios/Classes/methods/observers/SongsObserver.swift
class SongsObserver : NSObject, FlutterStreamHandler {
    
    private var obs: ObserverController?
    private var sink: FlutterEventSink?
    
    private var oldMedia: [[String: Any?]] = Array()
    
    private var pQuery: SongsQuery?
    private var pIsRunning: Bool = false
    
    var isRunning: Bool {
        get { pIsRunning }
    }
    
    var query: SongsQuery {
        get { pQuery! }
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        //
        sink = events
        
        //
        pQuery = SongsQuery(
            call: nil,
            result: nil,
            sink: self.sink,
            args: arguments as? [String: Any]
        )
        
        //
        obs = ObserverController(songsObserver: self)
        obs!.start()
        pIsRunning = true
        
        //
        pQuery?.querySongs()
        
        //
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        //
        pIsRunning = false
        
        //
        sink = nil
        pQuery = nil
        
        if obs != nil && obs!.isRunning {
            obs!.stop()
        }
        
        return nil
    }
}
